File: Algorithm/leetcode_1315.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_1315.py
# Author            : Andy
# Date              : 2022.09.08
# Last Modified Date: 2022.09.08
# Last Modified By  : Andy


class Solution:

    def sumEvenGrandparent(self, root):
        ans = 0

        def dfs(node):
            nonlocal ans

            if node is None:
                return 0

            if node.left is not None:
                if node.left.left is not None:
                    ans += node.left.left.val
                if node.left.right is not None:
                    ans += node.left.right.val

            if node.right is not None:
                if node.right.left is not None:
                    ans += node.right.left.val

                if node.right.right is not None:
                    ans += node.right.right.val

            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ans

File: Algorithm/leetcode_1315.go
/**
 * File              : leetcode_1315.go
 * Author            : Andy
 * Date              : 2022.09.08
 * Last Modified Date: 2022.09.08
 * Last Modified By  : Andy
 */

package sumEvenGrandparent

func dfs(node *TreeNode, ans *int){
    if node == nil {
        return
    }

    if node.Val % 2 == 0 {
        if node.Left != nil {
            if node.Left.Left != nil{
                *ans += node.Left.Left.Val
            }

            if node.Left.Right != nil {
                *ans += node.Left.Right.Val
            }
        }

        if node.Right != nil {
            if node.Right.Left != nil{
                *ans += node.Right.Left.Val
            }

            if node.Right.Right !+ nil {
                *ans += node.Right.Right.val
            }

        }
    }

    dfs(node.Left)
    dfs(node.Right)
}
func sumEvenGrandparent(root *TreeNode) int {
	ans := 0
    dfs(root, &ans)
    return ans
}
